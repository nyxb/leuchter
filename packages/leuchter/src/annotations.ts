import { annotateLine } from './annotations.inline'
import { annotateLines } from './annotations.multiline'
import type { Annotation } from './comments'
import type { InlineRange, MultiLineRange } from './range'

export interface Token {
   content: string
   style: {
      color?: string
      fontStyle?: 'italic'
      fontWeight?: 'bold'
      textDecoration?: 'underline' | 'line-through'
   }
   scopes?: string[]
}

export interface TokenGroup {
   annotationName: string
   annotationQuery?: string
   fromColumn: number
   toColumn: number
   tokens: Tokens
}

export type Tokens = (Token | TokenGroup)[]

export interface Line { lineNumber: number; tokens: Tokens }
export interface LineGroup {
   annotationName: string
   annotationQuery?: string
   fromLineNumber: number
   toLineNumber: number
   lines: Lines
}
export type Lines = (Line | LineGroup)[]

export function applyAnnotations(lines: Token[][], annotations: Annotation[]) {
   const { inlineAnnotations, multilineAnnotations }
    = splitAnnotations(annotations)

   const annotatedLines = lines.map((line, lineIndex) => {
      const lineAnnotations = inlineAnnotations.filter(
         annotation => annotation.range.lineNumber === lineIndex + 1,
      )
      return {
         lineNumber: lineIndex + 1,
         tokens: annotateLine(line, lineAnnotations),
      } as Line
   })

   return annotateLines(annotatedLines, multilineAnnotations)
}

export type InlineAnnotation = Omit<Annotation, 'ranges'> & {
   range: InlineRange
}

export type MultilineAnnotation = Omit<Annotation, 'ranges'> & {
   range: MultiLineRange
}

function splitAnnotations(annotations: Annotation[]) {
   const inlineAnnotations: InlineAnnotation[] = []
   const multilineAnnotations: MultilineAnnotation[] = []

   annotations.forEach((annotation) => {
      annotation.ranges.forEach((range) => {
         if ('lineNumber' in range)
            inlineAnnotations.push({ ...annotation, range })

         else
            multilineAnnotations.push({ ...annotation, range })
      })
   })

   return {
      inlineAnnotations,
      multilineAnnotations,
   }
}
