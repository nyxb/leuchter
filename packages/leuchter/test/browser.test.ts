import { extractAnnotations, highlight } from '../dist/browser.esm.mjs'
import { runTests } from './highlight.js'
import { runAnnotationTests } from './annotations.js'

runTests({ highlight })
runAnnotationTests({ extractAnnotations, highlight })
