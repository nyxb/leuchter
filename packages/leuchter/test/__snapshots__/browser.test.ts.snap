// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`extract annottations from text 1`] = `
{
  "annotations": [],
  "code": "const x = 1;
// foo[3:5]
const y = 2;",
}
`;
